### ISSUE TITLE
Azure Cognitive Search: SearchResourceEncryptionKey raises 'not iterable' error during data source connection creation in azure-search-documents SDK (v11.6.0b4))

### ISSUE THREAD

**Issue by YVass1** [Azure/azure-sdk-for-python#37251]
- **Package Name**:  azure.search.documents 
- **Package Version**:   11.6.0b4
- **Operating System**: OSX, version: 14.6.1 
- **Python Version**: 3.9.6

**Describe the bug**
When attempting to use the azure-search-documents SDK, the encryption key provided is not accepted due to a type mismatch.  
The SDK does not handle the type correctly, leading to a failure during deserialization when creating the data source connection. As according to the Microsoft documentation, SearchResourceEncryptionKey object is passed, but an error arises, where an iterable type is expected for encryption key. 

Microsoft docs: https://learn.microsoft.com/en-us/python/api/azure-search-documents/azure.search.documents.indexes.models.searchindexerdatasourceconnection?view=azure-python 
Python Library: https://pypi.org/project/azure-search-documents/11.5.0/ 

**To Reproduce**
Steps to reproduce the behaviour:
Assuming Blob Storage Account , AI Search, Index and indexer all already exists and managed identity enabled. 

1. Install the library azure-search-documents==11.6.0b4
2. Define SearchResourceEncryptionKey object for encryption.
3. Create Indexer client. 
4. Define data source connection passing SearchResourceEncryptionKey object defined earlier.
5. Attempt to create a data source connection for ab indexer.
6. Observe the "not iterable" error during the deserialization process.

```py
from azure.search.documents.indexes import SearchIndexerClient
from azure.core.credentials import AzureKeyCredential
from azure.search.documents.indexes.models import SearchIndexerDataSourceConnection, SearchResourceEncryptionKey

# Initialize the client
indexer_client = SearchIndexerClient(endpoint="https://<your-search-service>.search.windows.net",
                             credential=AzureKeyCredential("<your-key>"))

# Define an encryption key
encryption_key = SearchResourceEncryptionKey(
    key_name="your-key-name",
    key_version="your-key-version",
    vault_uri="https://<your-keyvault-name>.vault.azure.net/",
    application_id="<your-client-app-id>",
    application_secret="<your-server-app-secret>",
)

# Attempt to create a data source connection for an indexer
data_source_connection = SearchIndexerDataSourceConnection(
    name="my-datasource",
    type="azureblob",
    connection_string=<your-blob-managed-identity-connection-string>,
    container=SearchIndexerDataContainer(name=<your-container>, query=<your-folder-path>)
    encryption_key=encryption_key  # Pass SearchResourceEncryptionKey
)

try:
    indexer_client.create_or_update_data_source_connection(data_source_connection)
except Exception as e:
    print(f"Failed to create data source connection: {e}")
```

**Expected behaviour**
The SDK should correctly handle the SearchResourceEncryptionKey when creating a data source connection for an indexer. 

**Actual behaviour**
The operation fails with a "not iterable" error, suggesting that the SDK is not properly handling the SearchResourceEncryptionKey object during deserialization.

Error Message I receive:
```connections: ['[(<aiohttp.client_proto.ResponseHandler object at 0x13f74f820>, 144.713579458)]']
connector: <aiohttp.connector.TCPConnector object at 0x13f764340>
<class 'azure.search.documents.indexes.models._models.SearchResourceEncryptionKey'>
--- Logging error ---
Traceback (most recent call last):
  File "/Users/redacted-project-path/.venv/lib/python3.9/site-packages/azure/search/documents/indexes/_generated/_serialization.py", line 1461, in _deserialize
    found_value = key_extractor(attr, attr_desc, data)
  File "/Users/redacted-project-path/.venv/lib/python3.9/site-packages/azure/search/documents/indexes/_generated/_serialization.py", line 1230, in rest_key_case_insensitive_extractor
    return attribute_key_case_insensitive_extractor(key, None, working_data)
  File "/Users/redacted-project-path/.venv/lib/python3.9/site-packages/azure/search/documents/indexes/_generated/_serialization.py", line 1257, in attribute_key_case_insensitive_extractor
    for key in data:
TypeError: 'SearchResourceEncryptionKey' object is not iterable```

**Comment by YVass1** [comment]
Also tried with azure.search.documents package version: 11.6.0b4. Same error.

**Comment by kristapratico** [comment]
@YVass1 thanks for your detailed issue, the team will investigate and get back to you as soon as possible.

**Comment by xiangyan99** [comment]
Thanks for reaching out.

I tried your code snippet with azure-search-documents 11.5.1 but could not repo it.

Did you have a chance to create a new virtual env and try again?

**Comment by YVass1** [comment]
Hi, 

yes I recreated my venv and upgraded python version to 3.11 and package version to 11.6.0b4. The error still persists. I will try again with the combination python version 3.11 and package version 11.5.1

**Comment by xiangyan99** [comment]
Thanks for checking.

FYI. My testing env was Python 3.10.9 + azure-search-documents 11.5.1 on Windows.

**Comment by YVass1** [comment]
I recreated my virtual environment with 11.5.1 and python version 3.11. Note I am on OSX. Double checked library, version and path and showed path as expected in venv site packages (see screenshots below). But I receive this error:

<img width="869" alt="Screenshot 2024-09-10 at 21 57 06" src="https://github.com/user-attachments/assets/8396d5ef-c2bc-400b-be58-99c969b281e8">
<img width="816" alt="Screenshot 2024-09-10 at 21 56 57" src="https://github.com/user-attachments/assets/2d5ad945-961b-477c-b322-5afa75895fa5">


```
File "/Users/<redacted-project-path>/.venv/lib/python3.11/site-packages/azure/search/documents/indexes/_generated/aio/operations/_data_sources_operations.py", line 233, in create_or_update
    raise HttpResponseError(response=response, model=error)
azure.core.exceptions.HttpResponseError: () The request is invalid. Details: encryptionKey.keyVaultUri : The keyVaultUri field is required.
encryptionKey.keyVaultKeyName : The keyVaultKeyName field is required.
Code: 
Message: The request is invalid. Details: encryptionKey.keyVaultUri : The keyVaultUri field is required.
encryptionKey.keyVaultKeyName : The keyVaultKeyName field is required.
```
This seems to suggests the library either expects an initially encryption key object with attributes keyVaultUri and keyVaultKeyName or should have parsed the SearchResourceEncryptionKey input into the expected object, but this did not happen.

If I change the code so the params passed to SearchResourceEncryptionKey object match the error message, I receive this error. Same error if i try variation on the param names. : 

```
encryption_key = SearchResourceEncryptionKey(
        keyVaultUri="https://<my-key-vault-name>vault.azure.net/",
        keyVaultKeyName ="<my-key-name>",
        application_id="<my-app-id>",
        application_secret="<my-app-secret>",
        ) 
        
encryption_key = SearchResourceEncryptionKey(
        key_vault_uri="https://<my-key-vault-name>vault.azure.net/",
        key_vault_key_name ="<my-key-name>",
        application_id="<my-app-id>",
        application_secret="<my-app-secret>",
  ) 

```

```
Traceback (most recent call last):
  File "Users/<redacted-file-path>", line 174, in setup
    direct_encryption_key = SearchResourceEncryptionKey(
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: SearchResourceEncryptionKey.__init__() missing  --- Logging error ---
Traceback (most recent call last):
  File "Users/<redacted-file-path>", line 174, in setup
    encryption_key = SearchResourceEncryptionKey(
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: SearchResourceEncryptionKey.__init__() missing 3 required keyword-only arguments: 'key_name', 'key_version', and 'vault_uri'
```
But this is what i would expect as the library documentation does not say these are valid params. 

If I try to pass both variations shown in the code below, the other params are not recognised and skipped anyways and I have attribute error again:

```
encryption_key = SearchResourceEncryptionKey(
        key_name="<my-key-name>",
        key_version="<my-cmk-version>",
        vault_uri="https://<my-key-vault>.vault.azure.net/",
        keyVaultUri="https://<my-key-vault-name>vault.azure.net/",
        keyVaultKeyName ="<my-key-name>",
        application_id="<my-app-id>",
        application_secret="<my-app-secret>",
        )  
```
```
2024-09-10 21:03:05,272 - WARNING - azure.search.documents.indexes._generated._serialization - keyVaultUri is not a known attribute of class <class 'azure.search.documents.indexes.models._models.SearchResourceEncryptionKey'> and will be ignored
2024-09-10 21:03:05,272 - WARNING - azure.search.documents.indexes._generated._serialization - keyVaultKeyName is not a known attribute of class <class 'azure.search.documents.indexes.models._models.SearchResourceEncryptionKey'> and will be ignored
```

```
  File "/Users/<redacted-file-path>", line 233, in create_or_update
    raise HttpResponseError(response=response, model=error)
azure.core.exceptions.HttpResponseError: () The request is invalid. Details: encryptionKey.keyVaultUri : The keyVaultUri field is required.
encryptionKey.keyVaultKeyName : The keyVaultKeyName field is required.
Code: 
Message: The request is invalid. Details: encryptionKey.keyVaultUri : The keyVaultUri field is required.
encryptionKey.keyVaultKeyName : The keyVaultKeyName field is required.

```
I am not sure where else to look further or try as I have already recreated venv, checked path and library versions, and tried different combos. Though to me this suggests there is version mismatch between libraries that are being used perhaps. 
azure-search-documents uses/depends on the following libraries and here the versions installed for my project:
 - azure-core==1.30.2
 - azure-common==1.1.28
 - isodate==0.6.1

**Comment by xiangyan99** [comment]
The error:
File "/Users/<redacted-project-path>/.venv/lib/python3.11/site-packages/azure/search/documents/indexes/_generated/aio/operations/_data_sources_operations.py", line 233, in create_or_update
    raise HttpResponseError(response=response, model=error)
azure.core.exceptions.HttpResponseError: () The request is invalid. Details: encryptionKey.keyVaultUri : The keyVaultUri field is required.
encryptionKey.keyVaultKeyName : The keyVaultKeyName field is required.
Code: 
Message: The request is invalid. Details: encryptionKey.keyVaultUri : The keyVaultUri field is required.
encryptionKey.keyVaultKeyName : The keyVaultKeyName field is required.

meant the client successfully serialized the request and it was rejected by the service.

Feel free to open a ticket at https://azure.microsoft.com/support/create-ticket to ask some helps from the service side.

**Comment by YVass1** [comment]
Hi, 

Sorry can you please try to replicate the error with azure-search-documents package version 11..6.b4 as we further investigation found the original error is associated that version. While 11.5.1 also causes an error but a different one (for which you suggested to raise a support case - i have done so), moving forward we require package version 11.6.b4 as it contains required features. And can you please reopen this issue. Thank you.

**Comment by mattgotteiner** [comment]
Thanks for reporting this - I’ll try to reproduce this issue

**Comment by mattgotteiner** [comment]
I can reproduce this using 11.6.0b4

**Comment by YVass1** [comment]
Hi, @xiangyan99  Thanks for looking into the issue. Is it possible to have a new pre release for the package ?

**Comment by mattgotteiner** [comment]
@xiangyan99  what's the resolution to the issue? The issue was closed but AFAIK there's no new package yet

**Comment by YVass1** [comment]
@xiangyan99 
Hi, 

Is there any update on next step since the issue still persists and I ideally I would like access to a pre-release version so development can continue.

**Comment by xiangyan99** [comment]
We just shipped 11.6.0b5. https://pypi.org/project/azure-search-documents/11.6.0b5/

Please let us know if it does not work for you.

**Comment by YVass1** [comment]
Hi, thanks I can confirm this particular issue is resolved but there is another bug I have found I believe in relation to the `SearchIndexerSkillset` and `SearchResourceEncryptionKey `. Here is the link if anyone else in the same position: https://github.com/Azure/azure-sdk-for-python/issues/37521
